Q1

a. Dimension reduction
b. Clustering
c. Regression
d. Classification

**SVM-C with Gaussian kernel**

Target output type: one-hot

Parameters: support vectors, their coefficients, and intercepts

Hyperparameters: C (regularization), gamma (kernel coefficient), tol (tolerance)

Hyperparameter range: 0.5 to 2, 'scale' or 'auto', 1e-3 to 1e-5

sklearn commands: s=sklearn.svm.SVC(); s.fit(X, y); s.predict(X_)


**SVM-R with Gaussian kernel**

Target output type: float

Parameters: support vectors, their coefficients, and intercepts

Hyperparameters: C (regularization), gamma (kernel coefficient), tol (tolerance), epsilon

Hyperparameter range: 0.5 to 2, 'scale' or 'auto', 1e-3 to 1e-5, 0.1 to 0.01

sklearn commands: r=sklearn.svm.SVR(); r.fir(X, y); r.predict(X_)

**NN with one hidden layer (C)**

Target output type: integer(0/1)

Parameters: weight and bias matrices for each neuron

Hyperparameters: hidden_layer_size, activation, alpha(regularization)

Hyperparameter range: 20 to 500, 'relu' or 'logistic' or 'tanh', 0.0001 to 1

sklearn commands: clf = MLPClassifier(); clf.fit(X, y); clf.predict(X_)

**NN with one hidden layer (R)**

Target output type: float

Parameters: weight and bias matrices for each neuron

Hyperparameters: hidden_layer_size, activation, alpha(regularization)

Hyperparameter range: 20 to 500, 'relu' or 'logistic' or 'tanh', 0.0001 to 1

sklearn commands: clf = MLPRegressor(); clf.fit(X, y); clf.predict(X_)

Random forest (R)
Target output type: 
Parameters: 
Hyperparameters: 
Hyperparameter range: 
sklearn commands: 

Random forest (C)
Target output type: 
Parameters: 
Hyperparameters: 
Hyperparameter range: 
sklearn commands: 

k-means
Target output type: 
Parameters: 
Hyperparameters: 
Hyperparameter range: 
sklearn commands: 

DBSCAN
Target output type: 
Parameters: 
Hyperparameters: 
Hyperparameter range: 
sklearn commands: 

**PCA**

Target output type: none

Parameters: there are no specific parameters

Hyperparameters: num_components

Hyperparameter range: 1 to 3 for data visualization. Of the order of n/10 if some other framework has to be fit upon the data

sklearn commands: pca=sklearn.decomposition.PCA(n_components=2); X_train_ = pca.fit_transform(X_train); X_test_ = pca.transform(X_test);

Kernel PCA
Target output type: 
Parameters: 
Hyperparameters: 
Hyperparameter range: 
sklearn commands: 

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
df = pd.read_csv("/content/drive/My Drive/ds203_assignment/SeoulBikeData.csv", encoding='latin1')

In [ ]:
df.head()

Date  Rented Bike Count  Hour  ...  Seasons     Holiday  Functioning Day
0  01/12/2017                254     0  ...   Winter  No Holiday              Yes
1  01/12/2017                204     1  ...   Winter  No Holiday              Yes
2  01/12/2017                173     2  ...   Winter  No Holiday              Yes
3  01/12/2017                107     3  ...   Winter  No Holiday              Yes
4  01/12/2017                 78     4  ...   Winter  No Holiday              Yes

[5 rows x 14 columns]

In [ ]:
# this is supervised learning problem and a continuous(integer) variable has to be predicted
# performance will be measured by MSE
# we will drop date, as season helps account for it
# we will also drop dew point temperature as it is irrelevant
# all other variables (after apt preprocessing) will be used in the prediction
# i.e. all other variables are usable

In [ ]:
df.isnull().sum()

Date                         0
Rented Bike Count            0
Hour                         0
Temperature(°C)              0
Humidity(%)                  0
Wind speed (m/s)             0
Visibility (10m)             0
Dew point temperature(°C)    0
Solar Radiation (MJ/m2)      0
Rainfall(mm)                 0
Snowfall (cm)                0
Seasons                      0
Holiday                      0
Functioning Day              0
dtype: int64

In [ ]:
for col in df.columns:
  print(col, len(df[col].unique()))

Date 365
Rented Bike Count 2166
Hour 24
Temperature(°C) 546
Humidity(%) 90
Wind speed (m/s) 65
Visibility (10m) 1789
Dew point temperature(°C) 556
Solar Radiation (MJ/m2) 345
Rainfall(mm) 61
Snowfall (cm) 51
Seasons 4
Holiday 2
Functioning Day 2


In [ ]:
df["Holiday"]=df["Holiday"]=="Holiday"

In [ ]:
df["Functioning Day"]=df["Functioning Day"]=="Yes"

In [ ]:
for s in df["Seasons"].unique():
  df["is_"+s] = df["Seasons"]==s

In [ ]:
df=df.drop(["Date", "Seasons", "Dew point temperature(°C)"], axis=1)

In [ ]:
df.head()

Rented Bike Count  Hour  Temperature(°C)  ...  is_Spring  is_Summer  is_Autumn
0                254     0             -5.2  ...      False      False      False
1                204     1             -5.5  ...      False      False      False
2                173     2             -6.0  ...      False      False      False
3                107     3             -6.2  ...      False      False      False
4                 78     4             -6.0  ...      False      False      False

[5 rows x 15 columns]

In [ ]:
df_x = df.drop("Rented Bike Count", axis=1)
df_y=df["Rented Bike Count"]
data_x = df_x.to_numpy()
data_y = df_y.to_numpy()

In [ ]:
from sklearn.preprocessing import StandardScaler as SS
ss=SS()
data_x=ss.fit_transform(data_x)

In [ ]:
data_x

array([[-1.66132477, -1.51395724, -1.04248288, ..., -0.58051386,
        -0.58051386, -0.57629575],
       [-1.51686175, -1.53907415, -0.99336999, ..., -0.58051386,
        -0.58051386, -0.57629575],
       [-1.37239873, -1.58093567, -0.94425709, ..., -0.58051386,
        -0.58051386, -0.57629575],
       ...,
       [ 1.37239873, -0.86091752, -0.94425709, ..., -0.58051386,
        -0.58051386,  1.73522016],
       [ 1.51686175, -0.90277904, -0.8460313 , ..., -0.58051386,
        -0.58051386,  1.73522016],
       [ 1.66132477, -0.91952365, -0.74780551, ..., -0.58051386,
        -0.58051386,  1.73522016]])

In [ ]:
# framework1: Ridge Regression
# framework2: NN with one hidden layer

In [ ]:
l=data_x.shape[0]
indices = np.random.permutation(l)
idx_train, idx_val, idx_test = indices[:int(l*0.7)], indices[int(l*0.7):int(l*0.85)], indices[int(l*0.85):]
x_train, x_val, x_test = data_x[idx_train,:], data_x[idx_val,:], data_x[idx_test,:]
y_train, y_val, y_test = data_y[idx_train], data_y[idx_val], data_y[idx_test]

In [ ]:
from sklearn.linear_model import Ridge
from sklearn.neural_network import MLPRegressor

In [ ]:
ridge_alpha_vals=[1, 3, 10, 30, 100, 300, 1000]

In [ ]:
mse_min=10**10
a_min=0
for a in ridge_alpha_vals:
  r=Ridge(a, fit_intercept=True)
  r.fit(x_train, y_train)
  y_val_pred = r.predict(x_val)
  mse=np.sum((y_val-y_val_pred)**2)/y_val.shape[0]
  if mse<mse_min:
    mse_min=mse
    a_min=a
print(a_min, mse_min)

300 188375.6990703337


In [ ]:
nn_hl_sizes=[i*50 for i in range(2, 7)]
nn_alpha_vals=[0.01, 0.03, 0.1, 0.3, 1]

In [ ]:
mse_min=10**10
a_min=0
hl_min=0
lr_min=0
for hl_size in nn_hl_sizes:
  for a in nn_alpha_vals:
    nn = MLPRegressor(hidden_layer_sizes=(hl_size,), alpha=a, learning_rate_init=1, tol=1)
    # tolerance and lr_init have been changed for faster convergence
    nn.fit(x_train, y_train)
    y_val_pred = nn.predict(x_val)
    mse=np.sum((y_val-y_val_pred)**2)/y_val.shape[0]
    if mse<mse_min:
      mse_min=mse
      a_min=a
      hl_min=hl_size

In [ ]:
print(a_min, hl_min, mse_min)

0.03 150 58448.79569765229


One hidden layer Neural Network performs better

The optimal hyperparameters were found to be:

  alpha=0.03

  hidden_layer_size=150